Fix wrong results issue due to wrong nullability semantics of COUNT aggregate function

asenac opened the pull request:
This patch fixes an issue whereby any null rejecting filter on the result of a COUNT aggregation, resulted in the requirement for its parameter expression to be non-null.

Fixes #7049, #7047

wangandi commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -200,9 +200,17 @@ impl NonNullRequirements {
                     // A "non-empty" requirement, I guess?
```
I guess we should update this comment.

asenac commented on src/transform/src/column_knowledge.rs:
```diff
@@ -242,8 +242,12 @@ impl ColumnKnowledge {
                             // These methods propagate constant values exactly.
                             knowledge
                         }
+                        AggregateFunc::Count => DatumKnowledge {
```
This is need only for completeness. Both transforms overlap, ColumnKnowledge wasn't detecting `COUNT(..) is null` as a `false` predicate, but NonNullRequirements was.

wangandi replied:
Issue #6613 is about eliminating this overlap. I prefer eliminating the nonnull parts of `ColumnKnowledge` myself.

philip-stoev commented:
I am terribly terribly terribly sorry, in this branch this query has flipped from an outer-join to an inner join and the result is incorrect:

```
DROP SCHEMA PUBLIC cascade;
CREATE SCHEMA public;

CREATE TABLE t1 (f1 INTEGER, f2 INTEGER NOT NULL);
INSERT INTO t1 VALUES (NULL, 0);

CREATE TABLE t2 (f1 INTEGER, f2 INTEGER);
INSERT INTO t2 VALUES (NULL, 0);

CREATE TABLE t3 (f1 INTEGER, f2 INTEGER);
INSERT INTO t3 VALUES (NULL, 0);
INSERT INTO t3 VALUES (NULL, 0);

INSERT INTO t3 VALUES (1, 1);
INSERT INTO t3 VALUES (6, 6);

SELECT derived.agg1 FROM t1
JOIN (
	SELECT COUNT(*) AS agg1 , MAX(t2.f2) AS agg2
        FROM t2
        RIGHT JOIN t3 ON t3.f2 = 6
) AS derived ON TRUE
WHERE t1.f2 = derived.agg2;

 agg1 
------
    1 <- wrong, should have returned "4"
(1 row)
```

The plan in this branch is:

```
 %0 =                                        +
 | Get materialize.public.t1 (u246)          +
                                             +
 %1 =                                        +
 | Get materialize.public.t2 (u248)          +
 | ArrangeBy ()                              +
                                             +
 %2 =                                        +
 | Get materialize.public.t3 (u250)          +
 | Filter (#1 = 6)                           +
                                             +
 %3 =                                        +
 | Join %1 %2                                +
 | | implementation = Differential %2 %1.()  +
 | | demand = (#1)                           +
 | Reduce group=()                           +
 | | agg count(true)                         +
 | | agg max(#1)                             +
 | Filter !(isnull(#1))                      +
 | ArrangeBy (#1)                            +
                                             +
 %4 =                                        +
 | Join %0 %3 (= #1 #3)                      +
 | | implementation = Differential %0 %3.(#1)+
 | | demand = (#2)                           +
 | Project (#2)                              +
```

But the plan in the prior commit is:

```
 %0 =                                            +
 | Get materialize.public.t2 (u3969)             +
 | ArrangeBy ()                                  +
                                                 +
 %1 =                                            +
 | Get materialize.public.t3 (u3971)             +
 | Filter (#1 = 6)                               +
                                                 +
 %2 = Let l0 =                                   +
 | Join %0 %1                                    +
 | | implementation = Differential %1 %0.()      +
 | | demand = (#1, #2)                           +
                                                 +
 %3 =                                            +
 | Get materialize.public.t1 (u3967)             +
                                                 +
 %4 =                                            +
 | Get %2 (l0)                                   +
 | Distinct group=(#2)                           +
 | Negate                                        +
 | Map 6                                         +
                                                 +
 %5 =                                            +
 | Get materialize.public.t3 (u3971)             +
 | Distinct group=(#0, #1)                       +
                                                 +
 %6 =                                            +
 | Union %4 %5                                   +
                                                 +
 %7 =                                            +
 | Get materialize.public.t3 (u3971)             +
 | ArrangeBy (#0, #1)                            +
                                                 +
 %8 =                                            +
 | Join %6 %7 (= #0 #2) (= #1 #3)                +
 | | implementation = Differential %6 %7.(#0, #1)+
 | | demand = ()                                 +
 | Map null, null                                +
 | Project (#4, #5, #0, #1)                      +
                                                 +
 %9 =                                            +
 | Union %2 %8                                   +
 | Reduce group=()                               +
 | | agg count(true)                             +
 | | agg max(#1)                                 +
 | Filter !(isnull(#1))                          +
 | ArrangeBy (#1)                                +
                                                 +
 %10 =                                           +
 | Join %3 %9 (= #1 #3)                          +
 | | implementation = Differential %3 %9.(#1)    +
 | | demand = (#2)                               +
 | Project (#2)                                  +
```

asenac commented:
@philip-stoev, once again you saved me from committing something stupid :) Thanks a lot!

We can't discard null rows if there is at least one Count aggregation in the Reduction. With the current patch, I'm just discarding null rows in any of the columns from the non-preserving side is used by any non-count aggregation.

asenac commented:
So, this patch is not 100% correct I'm afraid. We can't trigger the outer-to-inner conversion  if there is at least a count aggregation

```
materialize=> select t1.f1, count(t1.f2), max(t2.f2) from t1 left join t2 on t1.f2 = t2.f1 group by t1.f1;
 f1 | count | max 
----+-------+-----
  1 |     3 |   2
(1 row)

materialize=> select t1.f1, count(t1.f2), max(t2.f2) from t1 left join t2 on t1.f2 = t2.f1 group by t1.f1 having max(t2.f2) > 0;
 f1 | count | max 
----+-------+-----
  1 |     1 |   2
(1 row)

materialize=> select * from t1;
 f1 | f2 
----+----
  1 |  0
  1 |  1
  1 |  1
(3 rows)

materialize=> select * from t2;
 f1 | f2 
----+----
  0 |  2
(1 row)

materialize=> 
```

asenac commented:
@wangandi, I've re-requested your review, since the final fix, that addresses the issues that Philip spotted, differs a lot from the original one I posted. Sorry about that.

philip-stoev commented:
Hey, Rafael Nadal, what about this one: With your latest commit:

```
commit 6a2adbf064a270756b7bcc77b280bf0b1584c755 (HEAD -> 7049-aggregation-null-requeriments, origin/7049-aggregation-null-requeriments)
Author: Andres Senac <asenac@materialize.com>
Date:   Wed Jun 16 11:24:25 2021 +0200

    Null rows cannot be discarded if there is any Count aggregation
```

The plan for this query is an inner join, but the result is incorrect. This branch returns NULL **for this particular column**, the number of rows is correct.

```
DROP SCHEMA public CASCADE;

CREATE SCHEMA public;

CREATE TABLE t1 (f1 INTEGER, f2 INTEGER);
INSERT INTO t1 VALUES (NULL, 0);
INSERT INTO t1 VALUES (1, 1);
INSERT INTO t1 VALUES (1, 1);
INSERT INTO t1 VALUES (2, 2);
INSERT INTO t1 VALUES (2, 2);

CREATE TABLE t2 (f1 INTEGER, f2 INTEGER);
INSERT INTO t2 VALUES (NULL, 0);
INSERT INTO t2 VALUES (NULL, 0);
INSERT INTO t2 VALUES (1, 1);

SELECT derived.agg1 FROM t2
JOIN (
	SELECT MIN(t2.f1) AS agg1, MAX(t1.f2) AS agg2
	FROM t1
	RIGHT JOIN t2
	ON t2.f2  + t1.f2  + t1.f1  = t1.f2  + t1.f1
) AS derived
ON t2.f1  < derived.agg2;
```

asenac commented:
Thanks, Novak! Once again, I got it wrong. Simplified version:

```
materialize=>  
 SELECT MIN(t2.f1) AS agg1, MAX(t1.f2) AS agg2
 FROM t1
 RIGHT JOIN t2
 ON t2.f2  + t1.f2  + t1.f1  = t1.f2  + t1.f1 having max(t1.f2) > 0;
 agg1 | agg2 
------+------
      |    2
(1 row)

materialize=>  
 SELECT MIN(t2.f1) AS agg1, MAX(t1.f2) AS agg2
 FROM t1
 RIGHT JOIN t2
 ON t2.f2  + t1.f2  + t1.f1  = t1.f2  + t1.f1;
 agg1 | agg2 
------+------
    1 |    2
(1 row)

materialize=> 
```

If there are multiple aggregations, imposing a non-null constraint on the input of one of them may alter the result of another one. I guess that's why the original code was only pushing down constraints from aggregation parameters only if there was a single aggregation.

Working on a fix.

frankmcsherry commented:
Random drive-by comment, but this seems like a great example where the SQL -> MIR transform could remove the magic associated with `NULL` and instead rewrite such aggregations as e.g. summing both normal values, and summing an indicator for `NULL`.

philip-stoev reviewed (approved):
All the wrong results are now gone, so we are good. No assertions.

wangandi commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,14 +195,55 @@ impl NonNullRequirements {
+                    // Note: all columns from the non-preserving side of an outer join could
+                    // be treated as a single column for this analysis, but we don't have
+                    // that information here.
+                    let new_nonnull_columns_per_aggregate = aggregates
```
for each aggregate, see which columns would be have nonnull if the aggregation is nonnull

wangandi commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,14 +195,55 @@ impl NonNullRequirements {
+                            aggr_nonnull_columns
+                        })
+                        .collect::<Vec<_>>();
+                    let all_aggregated_nonnull_columns = new_nonnull_columns_per_aggregate
```
Can `all_aggregated_nonnull_columns` be renamed to something to reflect the fact that it is an intersection of `new_nonnull_columns_per_aggregate`?

asenac replied:
I've called them `nonnull_req_per_aggregate` and `req_intersection`, where `req` stands for requirementes.

I've also added a sorter version of computing the intersection.

wangandi commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,14 +195,55 @@ impl NonNullRequirements {
-                    }
-                    if column == group_key.len() && aggregates.len() == 1 {
-                        aggregates[0].expr.non_null_requirements(&mut new_columns);
+                for column in columns.iter().filter(|c| **c < group_key.len()) {
```
Hmm what about using partition to split columns into group_key_columns vs aggs_columns? 
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition

asenac replied:
done

frankmcsherry commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,14 +195,52 @@ impl NonNullRequirements {
+                }
+
+                if !aggs_columns.is_empty() {
+                    // We can convert a non-null requirement on an aggregate function into
+                    // a non-null requirement on its parameter only if we are sure that
+                    // won't affect the result of any other aggregation.
```
Can you say more about this? I would have thought we want the union of the constraints. The intent of the analysis is that we can safely drop any row that exhibits a null value in any of the listed columns, in which case preserving the values of other columns seems irrelevant.

asenac replied:
We can't union constraints as the following example shows:

```
materialize=> select t1.f1, sum(t1.f2), max(t2.f2) from t1 left join t2 on t1.f2 = t2.f1 group by t1.f1 having max(t2.f2) >= 0;
 f1 | sum | max 
----+-------+-----
  1 |     2 |   2
(1 row)
materialize=> select * from t1 left join t2 on t1.f2 = t2.f1;
 f1 | f2 | f1 | f2 
----+----+----+----
  1 |  1 |    |   
  1 |  1 |    |   
  1 |  0 |  0 |  2
(3 rows)
materialize=> 
```

`max(t2.f2)` is required to be non-null, but we can't require `t2.f2` to be non-null, because that would result in `sum(t1.f2)` being 0, rather than 2 in that example, due to the null rows being removed.

We can push down the non-null requirement iff there is no other aggregation which result might be affected.

Instead of computing the union, we intersect the sets of columns that are required for the input of all the aggregations to be non-null. We can require the input columns of the non-null aggregations to be non-null iff no other aggregation may have a non-null input from a row with null values in all the columns in the non-null aggregation, ie. all the columns in the non-null aggregation must be present in that intersection of columns.

frankmcsherry replied:
I'll have to re-read that last paragraph a few times, but to double check our understanding: the intent of the transformation is that it is pushing down the information "this row will be dropped if it contains a `null` in any of these columns". Under that interpretation, I don't yet understand why we intersect the requirements.

There is some moment at which we *introduce* new requirements, for example with an empty non-null requirements, but a reduce who plans to ignore some rows with null columns. I agree that in that case we want to intersect the constraints, but it isn't what the text above in the comment says ("We can convert a non-null requirement on an aggregate function into a non-null requirement on its parameter ... ").

It looks like the code previously only triggered in the single-aggregation case, where intersection is a no-op; was it incorrect, or just missed some optimization?

asenac replied:
For fixing the wrong results issue this ticket is about, we just need the changes in `relation/mod.rs` and use that new `AggregateExpr::non_null_requirements` method in the old code for the single-aggregation case.

However, with this patch I'm also fixing the missing optimization opportunities, where all the aggregates are on a common set of columns. 

Sorry for the unclear comments. I'll try explaining them better :)

asenac replied:
> There is some moment at which we introduce new requirements, for example with an empty non-null requirements, but a reduce who plans to ignore some rows with null columns. I agree that in that case we want to intersect the constraints...

That is not what this code is doing because that wouldn't be correct in SQL, and that's why the comment doesn't say that, I think.

asenac replied:
The intent of the transformation is to push down the information "rows with null values in any of these columns won't change the result of the reduction and hence, can be discarded at optimization time"

frankmcsherry replied:
I think I understand what is going on now (or at least, retract my concern). What I meant wrt the above comment is that e.g. `count` is able to propose that some columns be added to the non-null requirements, even if the count output column is not itself required to be non-null, which is the sort of place these constraints come from in the first place. Most other aggregates *propagate* constraints. And once we have all of these in one place, we do want to intersect them, you are right.

asenac replied:
oh, you're right.

```
# outer join to inner join conversion allowed
query T multiline
EXPLAIN select t1.f1, count(t2.f2) from t1 LEFT JOIN t2 ON t1.f1 = t2.f1 group by t1.f1;
----
%0 =
| Get materialize.public.t1 (u1)
| Filter !(isnull(#0))
| ArrangeBy (#0)

%1 =
| Get materialize.public.t2 (u3)
| Filter !(isnull(#0))

%2 = Let l0 =
| Join %0 %1 (= #0 #2)
| | implementation = Differential %1 %0.(#0)
| | demand = (#0, #3)

%3 =
| Get materialize.public.t1 (u1)

%4 =
| Get %2 (l0)
| Distinct group=(#0)
| ArrangeBy (#0)

%5 =
| Join %3 %4 (= #0 #2)
| | implementation = Differential %3 %4.(#0)
| | demand = (#0)
| Negate
| Project (#0, #1)

%6 =
| Get materialize.public.t1 (u1)

%7 =
| Union %5 %6
| Map null, null

%8 =
| Union %7 %2
| Reduce group=(#0)
| | agg count(#3)

EOF

# outer join to inner join conversion allowed
query T multiline
EXPLAIN select t1.f1, count(t2.f2), max(t2.f2) from t1 LEFT JOIN t2 ON t1.f1 = t2.f1 group by t1.f1;
----
%0 =
| Get materialize.public.t1 (u1)
| Filter !(isnull(#0))
| ArrangeBy (#0)

%1 =
| Get materialize.public.t2 (u3)
| Filter !(isnull(#0))

%2 = Let l0 =
| Join %0 %1 (= #0 #2)
| | implementation = Differential %1 %0.(#0)
| | demand = (#0, #3)

%3 =
| Get materialize.public.t1 (u1)

%4 =
| Get %2 (l0)
| Distinct group=(#0)
| ArrangeBy (#0)

%5 =
| Join %3 %4 (= #0 #2)
| | implementation = Differential %3 %4.(#0)
| | demand = (#0)
| Negate
| Project (#0, #1)

%6 =
| Get materialize.public.t1 (u1)

%7 =
| Union %5 %6
| Map null, null

%8 =
| Union %7 %2
| Reduce group=(#0)
| | agg count(#3)
| | agg max(#3)

EOF

# outer join to inner join conversion allowed
query T multiline
EXPLAIN select t1.f1, count(t2.f2), max(t2.f2) from t1 LEFT JOIN t2 ON t1.f1 = t2.f1 group by t1.f1 having max(t2.f2) > 0;
----
%0 =
| Get materialize.public.t1 (u1)
| Filter !(isnull(#0))
| ArrangeBy (#0)

%1 =
| Get materialize.public.t2 (u3)
| Filter !(isnull(#0))

%2 = Let l0 =
| Join %0 %1 (= #0 #2)
| | implementation = Differential %1 %0.(#0)
| | demand = (#0, #3)

%3 =
| Get materialize.public.t1 (u1)

%4 =
| Get %2 (l0)
| Distinct group=(#0)
| ArrangeBy (#0)

%5 =
| Join %3 %4 (= #0 #2)
| | implementation = Differential %3 %4.(#0)
| | demand = (#0)
| Negate
| Project (#0, #1)

%6 =
| Get materialize.public.t1 (u1)

%7 =
| Union %5 %6
| Map null, null

%8 =
| Union %7 %2
| Reduce group=(#0)
| | agg count(#3)
| | agg max(#3)
| Filter (#2 > 0)

EOF
```

In those three cases we should be removing the outer join, but we aren't.

frankmcsherry replied:
I don't want to admit to knowing if the above does / doesn't check out, but another case I could imagine is 
```
SELECT
    max(col1),
    max(col2),
    max(col3),
FROM ..
```
whose columns don't intersect, but which can still remove outer joins. Though, not easily in the language of `nonnull_requirements`, which is communicating an "and" requirements on columns (any null can cause a drop) vs an "or" (if all columns are null, can drop).

It seems like maybe we should eventually roll this in to a predicate pushdown, which I think you've suggested before.

frankmcsherry replied:
btw, don't let all of these comments get in the way of 1. fixing bugs and 2. making strict improvements that may not capture everything.

philip-stoev commented:
In this branch, commit 751b03fa8, this query returns a wrong result due to outer-to-inner join conversion:

```
DROP SCHEMA public CASCADE;
CREATE SCHEMA public;

CREATE TABLE t1 (f1 INTEGER NOT NULL);
INSERT INTO t1 VALUES (1);

CREATE TABLE t2 (f1 INTEGER NOT NULL);
INSERT INTO t2 VALUES (1);
INSERT INTO t2 VALUES (2);

SELECT SUM(t1.f1 + t2.f1), SUM(t2.f1 + 0)
FROM t2
LEFT JOIN t1
ON t1.f1 < t2.f1
HAVING SUM(t1.f1 + t2.f1) > 0;

 sum | sum 
-----+-----
   3 |   2 <- incorrect, must be [3, 3]
```

I believe expressions in both aggregates and the HAVING are required to reproduce. The second aggregate can also be COUNT and the result is still incorrect. Not reproducible with the common-ancestor commit or on latest main.

asenac commented:
Thanks @philip-stoev! Sorry for yet another ping-pong match. I've just pushed a fix for that case.

philip-stoev commented:
All the other issues from the RQG runs today have been accounted for and none are specific to this branch. So I think this is good to go.

The number of new queries in the slt file alone is exemplary, a couple of PRs like that and our code coverage report will start showing > 100%.

frankmcsherry commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,58 @@ impl NonNullRequirements {
+                        })
+                        .collect::<Vec<_>>();
+                    let mut req_intersection: Option<HashSet<usize>> = None;
+                    for (nonnull_columns, nonnull_flag) in aggr_nonnull_req.iter() {
```
If I were writing this, I'd put (or like to think that I'd put) a comment like:
```
// Each aggregate may either 
// 1. directly support non-null pushdown, on any of an indicated set of columns, or
// 2. disqualify all other non-null pushdowns, as they require all records to be surfaced.
// <something about the relationship of `nonnull_flag` to this dichotomy.>
```

frankmcsherry commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,58 @@ impl NonNullRequirements {
+                        .collect::<Vec<_>>();
+                    let mut req_intersection: Option<HashSet<usize>> = None;
+                    for (nonnull_columns, nonnull_flag) in aggr_nonnull_req.iter() {
+                        if *nonnull_flag {
```
If this were something like `!is_disqualifying` or `should_push_down` that ends up being easier to locally verify, vs running up the code to sort out what the logical expression used there means.

frankmcsherry commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,58 @@ impl NonNullRequirements {
+                            } else {
+                                req_intersection = Some(nonnull_columns.clone());
+                            }
+                        } else if !aggr_nonnull_req
+                            .iter()
+                            .any(|(cols, flag)| *flag && cols.is_subset(&nonnull_columns))
```
I don't understand this guard. I would understand `else` and then clearing `req_intersection`, but it looks like there is a more thoughtful test going on here, I just can't read what its intent is.

asenac replied:
Hopefully it is more clear after the last pushed commit.

If, for example, MAX(column) is requested to be non-null, the presence of COUNT(column), or any other aggregate on `column`, should not prevent the non-null requirement on `column` from being pushed down. This is actually the new case I was trying to address with the last commit before the one I've just made.

frankmcsherry commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,58 @@ impl NonNullRequirements {
+                            (aggr_nonnull_columns, aggr_nonnull_flag)
+                        })
+                        .collect::<Vec<_>>();
+                    let mut req_intersection: Option<HashSet<usize>> = None;
```
Minor nit that `req_intersection` as a name describes what it is happening to the thing, rather than what the thing represents. I think what it means to represent is "candidate input columns whose nullness would allow dropping a row".

asenac replied:
I've renamed `req_intersection` as `pushable_nonnull_constraints`. Comment reads as follows

```
                        // If no aggregate prevents other non-null pushdowns, `pushable_nonnull_constraints`
                        // contains the columns whose nulls are ignored by all aggregates that can push down
                        // non-null constraints.

```

frankmcsherry reviewed (changes requested):
I struggled to read this top to bottom, which I think is a fine goal for code that has many subtleties. I left a bunch of notes about that. I hope their number isn't a problem; I just wanted to leave as much evidence as possible.

asenac commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,58 @@ impl NonNullRequirements {
+                    // propagate the non-null requirement as long as the requirements
+                    // of their parameters intersect with the requirements of a non-null
+                    // aggregate.
+                    let aggr_nonnull_req = aggregates
+                        .iter()
```
Actually it means the opposite. When the flag is true it means that the null filtering could be pushed down as long as it doesn't contradict the requirements for keeping the correct results of the rest of the aggregates.

asenac replied:
I've added that comment after adding the third case for the branch below that was hard to understand. Also, I've renamed the flag as `should_push_down`.

wangandi commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,80 @@ impl NonNullRequirements {
+                    for (should_push_down, ignores_nulls_on_columns) in
+                        nonnull_info_per_aggregate.iter()
+                    {
+                        // Each aggregate may either
```
Hmm perhaps this is better explained with examples?
* Two aggs: `sum(#0 + #2), max(#0 + #1)`, non-null requirements on both aggs should be pushed down => implies `!isnull(#0)`. We don't want to push down a `!isnull(#2)` because deleting a row like `(1,1, null)` would make the max wrong.
* Two aggs: `sum(#0 + #2), max(#0 + #1)`, non-null requirements on only the max should be pushed down  => ???

asenac replied:
actually, the second case implies !isnull(#0) I wasn't covering that case. Thanks!

asenac commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,80 @@ impl NonNullRequirements {
+                                        .is_subset(&ignores_nulls_on_columns)
+                            },
+                        ) {
+                            // 3. Early exist. The current aggregate discualifies all other non-null pushdowns as
```
yes, thanks!

asenac commented on src/expr/src/relation/func.rs:
```diff
@@ -600,6 +600,43 @@ impl AggregateFunc {
         scalar_type.nullable(nullable)
     }
+
+    /// Whether the result of the aggregate can be non-null with null inputs
```
Yes, exactly. The comment is wrong. It's a left-over from a previous version

asenac commented on src/transform/src/nonnull_requirements.rs:
```diff
@@ -195,16 +195,80 @@ impl NonNullRequirements {
+                    let nonnull_info_per_aggregate = aggregates
+                        .iter()
+                        .enumerate()
+                        .map(|(pos, aggr)| {
+                            let should_push_down = aggr.func.propagates_nonnull_constraint()
+                                && aggr_columns.contains(&(group_key.len() + pos));
+                            let mut ignores_nulls_on_columns = HashSet::new();
+                            if let repr::Datum::Null = aggr.func.identity_datum() {
+                                aggr.expr
+                                    .non_null_requirements(&mut ignores_nulls_on_columns);
+                            }
+                            (should_push_down, ignores_nulls_on_columns)
+                        })
+                        .collect::<Vec<_>>();
```
hmmm, I think that makes it more readable. Thanks!

asenac commented:
Hello reviewers! I've just pushed a commit addressing @wangandi's suggestions, which I think make the code way more readable. Thanks a lot, @wangandi!

wangandi reviewed (approved):
The current state does not look like it would result in us returning a wrong answer, so I'm ok with merging as is and leaving expanding the set of columns this applies to as a future TODO.

(Future expansion cases include:
`Sum(#1 + #0), Max(#0)`, nonnullrequirements only on the sum => implies `!isnull(#0), !isnull(#1)` )

asenac commented:
If @frankmcsherry agrees, I will merge this today and log a followup task for further expansions.
